ouroboros-network/src/Ouroboros/Network/NodeToNode.hs

```diff
@@ -538,3 +548,74 @@ localNetworkErrorPolicy = ErrorPolicies {
   where
     ourBug :: SuspendDecision DiffTime
     ourBug = Throw
+
+-- | Simple responder control function that discards the Protocol result.
+simpleResponderControl :: forall m a. (MonadSTM m)
+                       => (NodeToNodeProtocols -> MiniProtocolResponderControl m a)
+                       -> m ()
+simpleResponderControl rspFn = forever $
+    -- Fetch, but ignore the result of any miniprotocol so that it can
+    -- be restarted incase the initiator decides to talk to us again.
+    void $ atomically $ fetchResponderResult ChainSyncWithHeadersPtcl
+               `orElse` fetchResponderResult BlockFetchPtcl
+               `orElse` fetchResponderResult TxSubmissionPtcl
+  where
+    fetchResponderResult :: MonadSTM m => NodeToNodeProtocols -> STM m a
+    fetchResponderResult ptcl =
+        let (MiniProtocolResponderControl action) = rspFn ptcl in
+        action
+
+-- | Result of waiting on a list of miniprotocols to finish. Only used by simpleInitiatorControl
+data MiniProtocolClientResult = MpcRestart NodeToNodeProtocols
+                              | MpcResult  NodeToNodeProtocols
+
+-- | Simple initiator control function that starts all NodeToNodeProtocols and if one of them
+-- finishes it will wait for restartDelay seconds before restarting the protocol.
+simpleInitiatorControl :: forall m a.
+                          ( MonadAsync m
+                          , MonadSTM m
+                          , MonadTimer m
+                          )
+                       => DiffTime
+                       -> (NodeToNodeProtocols -> MiniProtocolInitiatorControl m a)
+                       -> m ()
+simpleInitiatorControl restartDelay ctrlFn = do
+    startQ <- atomically $ newTBQueue 3
+    atomically $ do
+        writeTBQueue startQ ChainSyncWithHeadersPtcl
+        writeTBQueue startQ BlockFetchPtcl
+        writeTBQueue startQ TxSubmissionPtcl
+    clientK startQ []
+
+  where
+    clientK :: TBQueue m NodeToNodeProtocols -> [(NodeToNodeProtocols, STM m a)] -> m ()
+    clientK restartQ resultActions = do
+        cr <- atomically $ waitOnRestartQueue restartQ
+                `orElse` foldr (orElse . waitOnMiniProtocolClientResult) retry resultActions
```

Comment:
I think you can turn this `foldr` into `asum . fmap waitOnMiniProcolClientResult`. `STM` has `Alternative` instance for which `<|>` is `orElse` and `empty` is `retry`.

Reply:
I wrote this without `<|>` and `Applicative` since it meant that I didn't have to force a constraint that was purely for ascetic reasons  on to the users of this function.

Reply:
@dcoutts I think it's worth adding `Alternative` to `MonadSTM` constraints:
```
class (Monad m, Monad (STM m), Alternative (STM m)) => MonadSTM m
```
Possibly also `MonadPlus`.

Reply:
I'm not going to address this. To me `foldr` and `orElse` is more readable.